### DISCUSSION TITLE
dynamic routes with server components

### DISCUSSION THREAD

**Discussion by lsaavedr** [makeswift/makeswift#782] [Ideas]
Hi, I'm making a template page to manage "products" by id, like in the Shopify example, the first thing is to create a provider:
```typescript
// app/product/provider.tsx
'use client'

import { createContext } from 'react'

export const ComplianceContext = createContext<string>('0')

export function ComplianceProvider({
  complianceId,
  children,
}: {
  complianceId: string
  children: React.ReactNode
}) {
  return <ComplianceContext.Provider value={complianceId}>{children}</ComplianceContext.Provider>
}
```
then use this provider in page route with the template:
```typescript
//app/product/[id]/page.tsx
import { notFound } from 'next/navigation'

import { Page as MakeswiftPage } from '@makeswift/runtime/next'
import { getSiteVersion } from '@makeswift/runtime/next/server'

import { ProductProvider } from '@/app/product/providers'
import { client } from '@/lib/makeswift/client'

type ParsedUrlQuery = { id: string }

export default async function Page({ params }: { params: ParsedUrlQuery }) {
  const snapshot = await client.getPageSnapshot('/__product__', {
    siteVersion: getSiteVersion(),
  })

  if (snapshot == null) return notFound()

  return (
    <ProductProvider productId={params.id}>
      <MakeswiftPage snapshot={snapshot} />
    </ProductProvider>
  )
}
```
Finally, a component is created that uses this provider, but any componen using a provider are a client component :(
I think that can be possible add an additional prop in `MakeswiftPage` passed to children components to make that in a server-component fashion, there exist plans to make some like that?, for example
```typescript
//app/product/[id]/page.tsx
import { notFound } from 'next/navigation'

import { Page as MakeswiftPage } from '@makeswift/runtime/next'
import { getSiteVersion } from '@makeswift/runtime/next/server'

import { client } from '@/lib/makeswift/client'

type ParsedUrlQuery = { id: string }

export default async function Page({ params }: { params: ParsedUrlQuery }) {
  const snapshot = await client.getPageSnapshot('/__product__', { // template page
    siteVersion: getSiteVersion(),
  })

  if (snapshot == null) return notFound()

  return (
      <MakeswiftPage snapshot={snapshot} componentProps={{productId: params.id}}/>
  )
}
```

**Comment by lsaavedr** [comment]
`searchParams` should be another `componentProps`

**Comment by migueloller** [comment]
The `MakeswiftPage` component is a client component so we can't pass props to a server component. Furthermore, Makeswift doesn't support registering server components right now (i.e., you can't do `runtime.registerComponent(AServerComponent, {})`)

Regardless, even if we supported registering server components, the idea of server context is one that's independent of Makeswift and is still an area of research in React. If you wanted to be able to access a value like this in a server component you'd have to use something like `react-cache` to get a faux server context that's unique per request. You could also use AsyncLocalStorage. Some APIs that implement these strategies are Next.js' `headers()` and `cookies()` APIs and [Apollo Client's experimental RSC support](https://github.com/apollographql/apollo-client-nextjs), in particular the `getClient` helper.

https://github.com/apollographql/apollo-client-nextjs/blob/main/packages/client-react-streaming/src/registerApolloClient.tsx#L97-L145
